### PR TITLE
Support revision and refactor storage to index + db

### PR DESCRIPTION
**Pull request by pzheng1025**

### CODE REVIEW

File: xline/src/storage/kvstore.rs [hunk 1]
```diff
@@ -35,8 +38,74 @@ pub(crate) struct KvStore {
...
 #[derive(Debug)]
 struct KvStoreInner {
-    /// B-tree map
-    btree: Mutex<BTreeMap<Vec<u8>, KeyValue>>,
+    /// Key Index
+    index: Mutex<BTreeMap<Vec<u8>, Vec<KeyRevision>>>,
+    /// DB to store key value
+    db: Mutex<HashMap<Revision, KeyValue>>,
+    /// Revision
+    revision: Mutex<i64>,
```
Comment (rogercloud) [inline xline/src/storage/kvstore.rs:46]: Is atomic faster for `i64` type?

Comment (rogercloud) [inline xline/src/storage/kvstore.rs:46]: Is atomic faster for `i64` type?

File: xline/src/storage/kvstore.rs [hunk 2]
```diff
@@ -35,8 +38,74 @@ pub(crate) struct KvStore {
 /// KV store inner
 #[derive(Debug)]
 struct KvStoreInner {
-    /// B-tree map
-    btree: Mutex<BTreeMap<Vec<u8>, KeyValue>>,
+    /// Key Index
+    index: Mutex<BTreeMap<Vec<u8>, Vec<KeyRevision>>>,
+    /// DB to store key value
+    db: Mutex<HashMap<Revision, KeyValue>>,
```
Comment (rogercloud) [inline xline/src/storage/kvstore.rs:44]: Maybe BtreeMap is better? We'll remove the early revisions for GC.

File: xline/src/storage/kvstore.rs [hunk 3]
```diff
@@ -35,8 +38,74 @@ pub(crate) struct KvStore {
...
+#[derive(Debug, Copy, Clone)]
+struct KeyRevision {
+    /// Last creation revision
+    create_revision: i64,
+    /// Number of modification since last creation
+    version: i64,
+    /// Last modification revision
+    mod_revision: i64,
+    /// Sub revision in one transaction
+    sub_revision: i64,
```
Comment (rogercloud) [inline xline/src/storage/kvstore.rs:59]: Do different sub revisions mean different key revision? If yes, a transaction may take many revisions. Is it what we're expecting?

File: xline/src/storage/kvstore.rs [hunk 4]
```diff
@@ -160,55 +225,111 @@ impl KvStoreInner {
...
+
     #[allow(clippy::pattern_type_mismatch)]
     /// Handle `RangeRequest`
-    fn handle_range_request(
-        btree: &MutexGuard<BTreeMap<Vec<u8>, KeyValue>>,
-        req: RangeRequest,
-    ) -> RangeResponse {
+    fn handle_range_request(&self, req: RangeRequest) -> RangeResponse {
+        let key_index = self.index.lock();
+        let db = self.db.lock();
```
Comment (rogercloud) [inline xline/src/storage/kvstore.rs:237]: If they're always locked at the same time, shall we move them into one lock?

Comment (rogercloud) [reply xline/src/storage/kvstore.rs:237]: For several locks acquired at the same time, we'd better document the order somewhere to make sure we keep the order everywhere.

File: xline/src/storage/kvstore.rs [hunk 5]
```diff
@@ -217,51 +338,83 @@ impl KvStoreInner {
...
             ONE_KEY => {
-                if let Some(kv) = btree.remove(key) {
-                    prev_kvs.push(kv);
-                    response.deleted = 1;
+                if let Some(indexes) = key_index.get_mut(key) {
+                    if let Some(index) = Self::get_index(indexes) {
+                        let prev_kv = db
+                            .get(&index.as_revision())
+                            .cloned()
+                            .unwrap_or_else(|| panic!("Failed to get key {:?} from DB", key));
```
Comment (rogercloud) [inline xline/src/storage/kvstore.rs:358]: Please also log the index or revision.

File: xline/src/storage/kvstore.rs (hunk 3 above)
Comment (pzheng1025) [reply xline/src/storage/kvstore.rs:59]: Operations in on txn will have the same global revision but different sub_revision. It doesn't mean different global revision.

File: xline/src/storage/kvstore.rs [hunk 6]
```diff
@@ -160,55 +225,111 @@ impl KvStoreInner {
...
+                let mut values: Vec<KeyValue> = key_index
+                    .values()
+                    .filter_map(|indexes| {
+                        Self::get_index(indexes)
+                            .and_then(|index| db.get(&index.as_revision()).cloned())
+                    })
+                    .collect();
                 kvs.append(&mut values);
             }
             _ => {
```
[the remark this answers is not in the document]
Comment (pzheng1025) [reply xline/src/storage/kvstore.rs:264]: This match is to match vector. It is impossible to list all others.

File: xline/src/storage/kvstore.rs (hunk 1 above)
Comment (pzheng1025) [reply xline/src/storage/kvstore.rs:46]: So far there is only one thread. use mutex is cheap when no contention.

File: xline/src/storage/kvstore.rs (hunk 4 above)
Comment (pzheng1025) [reply xline/src/storage/kvstore.rs:237]: I will separate index and db in to different module in next patch. Lock will be acquired in their own module.

File: xline/src/storage/kvstore.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (pzheng1025) [reply xline/src/storage/kvstore.rs]: Done

**Review by rogercloud — APPROVED** [review]

**Pull request merged**